### PR TITLE
ballle98/AqualinkD#13: add time sync

### LINKED ISSUE #93: PDA time sync

### PR DESCRIPTION
**Pull request by ballle98**
fix #93

### CODE REVIEW

File: pda_aq_programmer.c [hunk 1]
```diff
@@ -858,14 +852,46 @@ bool set_PDA_aqualink_freezeprotect_setpoint(struct aqualinkdata *aq_data, int v
...
+  } else if (! set_PDA_numeric_field_value(aq_data, val, aq_data->frz_protect_set_point, NULL, 1)) {
     logMessage(LOG_ERR, "Error failed to set freeze protect temp value\n");
     return false;
   } else {
       return waitForPDAnextMenu(aq_data);
   }
 }
 
+bool set_PDA_aqualink_time(struct aqualinkdata *aq_data) {
+  if (! goto_pda_menu(aq_data, PM_SET_TIME)) {
```
Comment (ballle98) [inline pda_aq_programmer.c]: going to SET_TIME menu will update aq_data->tm. And every time the menu is updated the tm value is also updated

**Comment by sfeakes** [comment]
I'll go over the changes what I get some time.  But previous implementation was inefficient, that's why I asked for changes and for it to fit within the current code structure.
Remember you get time as a string, pass it to the UI as a string, and update it as a string, and there is no reason to check it every time you get a new time.  So using the currently implemented logic, the time string only get's converted to time datatype every hour or so when you need to check the time is in-sync (might be every 10 mins, I can't remember).  But at that check is the only time you actually need to convert the time string.

File: pda_aq_programmer.c [hunk 2]
```diff
@@ -966,132 +962,41 @@ Debug:   PDA:       PDA Menu Line 8 = Press SELECT
...
-    send_cmd(KEY_PDA_SELECT);
-    send_cmd(KEY_PDA_BACK);
-    send_cmd(KEY_PDA_BACK);
-
-  } else {
-    LOG(PDA_LOG,LOG_ERR, "Error set time failed\n");
-    send_cmd(KEY_PDA_BACK);
-    send_cmd(KEY_PDA_BACK);
-    return true;
+  if (strptime(pda_m_line(2), "%t%D %a", &panel_tm) == NULL) {
```
Comment (ballle98) [inline pda_aq_programmer.c:965]: removed tm from aq_data now getting panel time from menu buffer

**Comment by ballle98** [comment]
@sfeakes I have rebased and updated this pull request to the latest version.

I have tested this thoroghly with my PDA pannel and it synce the time correctly.

I have granted permision to for you to modify this pull requests if you want to change anything before merging
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

There are instructions for working with the pull offline on a branch again you can make changes in additional commits before merging
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally?tool=webui

attached logs showing it working

[stdout-20230528_2203.txt](https://github.com/sfeakes/AqualinkD/files/11587905/stdout-20230528_2203.txt)

File: pda.c [hunk 3]
```diff
@@ -307,18 +307,20 @@ void process_pda_packet_msg_long_time(const char *msg)
...
   //         "     SAT 10:29AM"
   //         "     SAT 4:23PM "
+  //         "     SUN  2:36PM"
   // printf("TIME = '%.*s'\n",AQ_MSGLEN,msg );
   // printf("TIME = '%c'\n",msg[AQ_MSGLEN-1] );
+
   if (msg[AQ_MSGLEN - 1] == ' ')
   {
-    strncpy(_aqualink_data->time, msg + 9, 6);
+    snprintf(_aqualink_data->time, sizeof(_aqualink_data->time), "%.6s", msg + 9);
```
Comment (ballle98) [inline pda.c:316]: strncpy is an unsafe string function.  it assumes the src is null terminated

**Pull request merged**